plugin/pkg/proxy/connect.go

```diff
@@ -121,9 +121,24 @@ func (p *Proxy) Connect(ctx context.Context, state request.Request, opts Options
 			// For UDP, if the error is not a network error keep waiting for a valid response to prevent malformed
 			// spoofs from blocking the upstream response.
 			// In the case this is a legitimate malformed response from the upstream, this will result in a timeout.
-			if proto == "udp" {
-				if _, ok := err.(net.Error); !ok {
-					continue
+
+			// If the error is an overflow, we probably have an upstream misbehaving in some way.
+			// (e.g. sending >512 byte UDP responses without an eDNS0 OPT RR).
+			// Instead of returning an error, return an empty response with TC bit set. This will make the
+			// client retry over TCP (if that's supported) or at least receive a clean
+			// error. The connection is still good so we break before the close.
+			if proto == "udp" && strings.Contains(err.Error(), "overflow") {
```

Comment:
I had a question regarding if we should create types for these errors, searching for the string "overflow" seems a bit dangerous to me: https://github.com/coredns/coredns/pull/6003/files#r1156333895

Reply:
I have create an issue here - https://github.com/miekg/dns/issues/1463 and will have to wait before I can update this code - if proto == "udp" && ((strings.Contains(err.Error(), "overflow")) || dnsErrBufOccured)

Reply:
Created a PR for this - https://github.com/miekg/dns/pull/1465

Reply:
Thanks!  In that PR, maybe put a blank line between `ErrTime`  and `ErrOverflowUnpackingA`, that way indentation doesn't change on existing set of errors, and the diff looks cleaner.

Reply:
@chrisohaver , @gcs278 - While waiting for this PR https://github.com/miekg/dns/pull/1465 to be reviewed, I wanted to check with you if I need to check for all the overflow errors exported here - https://github.com/miekg/dns/pull/1465/files. Because Overflow error due to 512 cache size can occur due to any of the reasons.

For example - 

if errors.Is(err, dns.ErrOverflowHeaderSize)
if errors.Is(err, dns.ErrOverflowPackingUint8)
if errors.Is(err, dns.ErrBuf) .... so on

Reply:
> I had a question regarding if we should create types for these errors, searching for the string "overflow" seems a bit dangerous to me: https://github.com/coredns/coredns/pull/6003/files#r1156333895

@miekg - What would you recommend in this case ? Reference PR - https://github.com/miekg/dns/pull/1465

Reply:
@miekg - Any suggestions will be appreciated. I will look forward to hearing from you.

Reply:
@chrisohaver - Do you have any suggestions here ?

Reply:
@chrisohaver @miekg - I kindly request your review and suggest on what we can do here to move forward on this change. We need to fix this to move forward with upcoming releases.

Reply:
Miek is on indiefinate sabbatical from the CoreDNS project, for about a year now.

Looks like https://github.com/miekg/dns/pull/1465 was rejected.

Based on the [comment](https://github.com/miekg/dns/pull/1465#issuecomment-1584116670), I would suggest re-submitting it with following changes:

Don't create new class variables for every possible overflow case.
Create a single public overflow error type that implements error, and raise this error for overflows.
You can make the error text different per overflow case, but the error type would be the same for all.

Reply:
Thank you @chrisohaver - I have submitted a PR with the changes here - https://github.com/miekg/dns/pull/1475, waiting for it to be reviewed.